### PR TITLE
Parallelization and better date filtering

### PR DESCRIPTION
**Pull request by GalOshri**
Parallelization: download joined logs and models using parallel threads
Date filtering: instead of looking through all blobs and only downloading the blobs in the correct date range, look through only blobs in the correct date range.

Note: using "if __name__ == '__main__':" to avoid multithreading issues.

### CODE REVIEW

File: DataScience/DataScience.py [hunk 1]
```diff
@@ -48,25 +45,21 @@ def read(self, eventid):
...
-        # continue to read one more event to close the file
+            
+        # read all events in file
         ret = None
         for line in self.file:
             js = json.loads(line)
             js_event_id = js['_eventid']
             if (js_event_id == eventid):
-                ret = [line, js]
+                ret = line
```
Comment (eisber) [inline DataScience/DataScience.py]: why don't you keep the parsed JSON around?

File: DataScience/DataScience.py [hunk 2]
```diff
@@ -176,56 +132,121 @@ class CheckpointedModel:
...
-                ts = datetime(int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4)), int(m.group(5)), int(m.group(6)))  
-                if ts.date() < start_date or ts.date() > end_date:  
-                    continue  
-                yield CheckpointedModel(ts, 'onlinetrainer', time_container.name)
+                ts = datetime(int(m.group(1)), int(m.group(2)), int(m.group(3)), int(m.group(4)), int(m.group(5)), int(m.group(6)))    
+                yield (ts, 'onlinetrainer', time_container.name)
 
-model_history = list(get_checkpoint_models())
-model_history.sort(key=lambda jd: jd.ts)
+if __name__ == '__main__':
```
Comment (eisber) [inline DataScience/DataScience.py]: am I missing the caller code? Where is the parallelization done?

File: DataScience/DataScience.py [hunk 3]
```diff
@@ -79,12 +72,13 @@ def __init__(self, ts, blob):
         self.data = []
 
     def index(self, idx):
-        f = open(self.filename, 'r')
+        f = open(self.filename, 'r', encoding="utf8")
```
Comment (eisber) [inline DataScience/DataScience.py]: when do you use ' vs. "?

File: DataScience/DataScience.py (hunk 1 above)
Comment (GalOshri) [inline DataScience/DataScience.py]: It's not being used anywhere and it led to a bug (the pop in the first line of the function was returning just the line.

File: DataScience/DataScience.py (hunk 2 above)
Comment (GalOshri) [inline DataScience/DataScience.py]: The parallelization is done in 
"with Pool(processes = 5) as p:
       data = p.map(download_data, joined)"

File: DataScience/DataScience.py (hunk 3 above)
Comment (GalOshri) [inline DataScience/DataScience.py]: Changed it to '.

**Comment by eisber** [comment]
thanks!

**Pull request merged**